pyproject.toml

```diff
@@ -99,6 +99,10 @@ doc = [
 dev = [
     "mypy",
     "typing_extensions",
+    "sphinx",
+    "pybind11",
+    "pytest",
```

Comment:
Isn't `pybind11` already covered more appropriately in the build system section at the top? It is a build-time only dependency after all.

`pytest` is already covered in the `test` section above, and same for `sphinx` in the `doc` section. Perhaps you intend to "require" a newer sphinx if the type hints are recent? I know for `pytest` the type hints were very recent in `7.0.0`, and for testing deeps bleeding edge is usually just fine for us I think.

Reply:
> Perhaps you intend to "require" a newer sphinx if the type hints are recent?

This is basically the point, yes. More specifically: make sure that they're available in the first place before running mypy in the [linux.yml](https://github.com/scipy/scipy/blob/cf0696dea31c567fa07cc4fc5b74cba23dbc254d/.github/workflows/linux.yml#L90-L92) github actions workflow, as this didn't happen previously. 

I do agree that this introduces some duplication among the `optional-dependencies` blocks, which is admittedly less than ideal. An alternative would be to install them manually in aforementioned .yml file, which by the looks of it already happens for a few dependencies. I'm very much open to suggestions here, would there be a preference for either approach?

https://github.com/scipy/scipy/blob/cf0696dea31c567fa07cc4fc5b74cba23dbc254d/.github/workflows/linux.yml#L82-L88

Reply:
I have a preference for installing this in `linux.yml`, rather than adding/duplicating them in `pyproject.toml`. That's cleaner, and I think in the right place - `pyproject.toml` is meant more for "build from sdist" than for dev dependencies.

Reply:
The rest of this PR looks great to me.